### PR TITLE
Fix exact match in wildcard

### LINKED ISSUE #3443: Generated WildcardRegex will match substrigs of non registered sources.
## Symptom

When adding one activity source with wildcard the [TracerProviderSdk](https://github.com/open-telemetry/opentelemetry-dotnet/blob/f471a9f197d797015123fe95d3e12b6abf8e1f5f/src/OpenTelemetry/Trace/TracerProviderSdk.cs#L219) will build a regex for all the registered activity sources.

Having the activities: `CustomActivity` and `CustomWildcardActivity.*` it will generate the following regex: 
```
^(?:CustomActivity)|(?:CustomWildcardActivity\..*.)$
```

The problem is that this regex will match also activities like `CustomActivity.AnyString` which was not registered and `CustomActivity` dose not contain wildcard.

**What is the expected behavior?**

Sub activities names like `CustomActivity.AnyString` should not be exported.

**What is the actual behavior?**

Sub activities `CustomActivity.Validation` are exported.

## Reproduce
- register the following sources
```csharp
service.AddOpenTelemetryTracing(tracer =>
{
    tracer
        .AddSource("CustomActivity")
        .AddSource("CustomWildcardActivity.*")
        .AddConsoleExporter();
});
```
- generate some activities with `CustomActivity.AnyString`

### PR DESCRIPTION
**Pull request by glucaci**
Fixes #3443.

### CODE REVIEW

File: src/OpenTelemetry/Internal/WildcardHelper.cs [hunk 1]
```diff
@@ -41,7 +41,17 @@ public static Regex GetWildcardRegex(IEnumerable<string> patterns = default)
 
         var convertedPattern = string.Join(
             "|",
-            from p in patterns select "(?:" + Regex.Escape(p).Replace("\\*", ".*").Replace("\\?", ".") + ')');
+            from p in patterns select "(?:" + ReplacePatterns(p) + ')');
         return new Regex('^' + convertedPattern + '$', RegexOptions.Compiled | RegexOptions.IgnoreCase);
```
Comment (reyang) [inline src/OpenTelemetry/Internal/WildcardHelper.cs:45]: I think the fix should be this one line change:

```suggestion
        return new Regex("^(" + convertedPattern + ")$", RegexOptions.Compiled | RegexOptions.IgnoreCase);
```

**Review by reyang — CHANGES_REQUESTED** [review]
https://github.com/open-telemetry/opentelemetry-dotnet/pull/3446/files#r921378089

File: src/OpenTelemetry/Internal/WildcardHelper.cs (hunk 1 above)
Comment (reyang) [reply src/OpenTelemetry/Internal/WildcardHelper.cs:45]: Or maybe:

```suggestion
        return new Regex("^(?:" + convertedPattern + ")$", RegexOptions.Compiled | RegexOptions.IgnoreCase);
```

@Yun-Ting what's your suggestion?

Comment (Yun-Ting) [reply src/OpenTelemetry/Internal/WildcardHelper.cs]: Non-capturing group should be sufficient here.

File: test/OpenTelemetry.Tests/Internal/WildcardHelperTests.cs [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+// limitations under the License.
+// </copyright>
+
+using Xunit;
+
+namespace OpenTelemetry.Internal.Tests
+{
+    public class WildcardHelperTests
+    {
+        [Theory]
```
[the remark this answers is not in the document]
Comment (glucaci) [reply test/OpenTelemetry.Tests/Internal/WildcardHelperTests.cs:23]: Sure, done

**Review by reyang — APPROVED** [review]
LGTM with a changelog entry.

**Review by Yun-Ting — APPROVED** [review]

**Comment by glucaci** [comment]
Please let me know if the changelog entry is good enough to understand the issue which was fixed.

**Pull request merged**